judge/views/user.py

```diff
@@ -507,3 +512,98 @@ def post(self, request, *args, **kwargs):
+            'expiry_minutes': settings.DMOJ_EMAIL_CHANGE_EXPIRY_MINUTES,
+            'user': self.request.user,
+            'activation_key': activation_key,
+            'new_email': new_email,
+        }
+        send_mail(
+            self.notify_subject_template_name, self.notify_email_template_name, context, None, self.request.user.email,
+            self.notify_html_email_template_name,
+        )
+        send_mail(
+            self.activate_subject_template_name, self.activate_email_template_name, context, None, new_email,
+            self.activate_html_email_template_name,
+        )
+
+        return generic_message(
+            self.request,
+            _('Email change requested'),
+            _('Please click on the link sent to the new email address.'),
+        )
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        kwargs['user'] = self.request.user
+        return kwargs
+
+    def post(self, request, *args, **kwargs):
+        key = f'emailchange!{request.META["REMOTE_ADDR"]}'
+        cache.add(key, 0, timeout=settings.DMOJ_EMAIL_CHANGE_LIMIT_WINDOW)
+        if cache.incr(key) > settings.DMOJ_EMAIL_CHANGE_LIMIT_COUNT:
+            return HttpResponse(_('You have sent too many email change requests. Please try again later.'),
+                                content_type='text/plain', status=429)
+        return super().post(request, *args, **kwargs)
+
+
+class EmailChangeActivateView(LoginRequiredMixin, View):
+    def get(self, request, *args, **kwargs):
+        activation_key = kwargs['activation_key']
+        signer = signing.TimestampSigner()
+        try:
+            try:
+                data = signer.unsign_object(
+                    base64.urlsafe_b64decode(activation_key.encode()).decode(),
+                    max_age=settings.DMOJ_EMAIL_CHANGE_EXPIRY_MINUTES * 60,
+                )
+            except (binascii.Error, signing.BadSignature):
+                raise ValueError(_('Invalid activation key. Please try again.'))
+            except signing.SignatureExpired:
+                raise ValueError(_('This request is expired. Please try again.'))
+            if data['id'] != request.user.id:
+                raise ValueError(
+                    _('Please try again from the account this email change was originally requested from.'),
+                )
+            with revisions.create_revision(atomic=True):
+                if User.objects.filter(email=data['email']).exists():
+                    raise ValueError(_('The email originally requested is in use. Please try again with a new email.'))
+                request.user.email = data['email']
+                request.user.save()
+                revisions.set_user(request.user)
+                revisions.set_comment(_('Changed email address'))
+        except ValueError as ve:
+            return generic_message(request, _('Email change failed'), str(ve), status=403)
```

Comment:
When does this happen?

Reply:
Any of the above errors that get raised are caught here and returned as a generic message.

Reply:
Do we have this pattern elsewhere?

I'm tempted to say we should move the logic of changing to its own function, and make the displaying be in `get()`.

Also, is there a method that makes more sense than `str(ve)`

Reply:
Finally can we use a dedicated exception instead of `ValueError`? I thought we failed some conversion or smth.

Reply:
Changed to a `EmailChangedError`. I think `str(e)` is the standard way to retrieve the message from an error, unless I am mistaken.

Reply:
Are you against moving the logic into its own function to avoid the double try?

Reply:
I don't really see the point, since it'll only ever be used once, but I'll move it to a separate function if it makes it easier to read the code.

Reply:
Other people can comment on this.. I could live with the current code.

Reply:
I've moved it to a separate function.